common/interpolation/src/spline_interpolation.cpp

```diff
@@ -157,13 +235,142 @@ std::vector<double> slerp(
+// }
+
+// TODO(murooka) implement slerpYaw, just interpolates yaw directly from yaw in points.
+// NOTE: assuming that points have valid yaw information.
+// template <typename T>
+// std::vector<double> slerpYaw(const std::vector<T> & points);
+
+// template <typename T>
+// std::vector<double> slerpYawFromPoints(const std::vector<T> & points)
+std::vector<double> slerpYawFromPoints(const std::vector<geometry_msgs::msg::Point> & points)
+{
+  const auto base = getBaseValues(points);
+
+  const auto & base_s = base.at(0);
+  const auto & base_x = base.at(1);
+  const auto & base_y = base.at(2);
+
+  // calculate spline coefficients
+  const auto multi_spline_coef_x = getSplineCoefficients(base_s, base_x);
+  const auto multi_spline_coef_y = getSplineCoefficients(base_s, base_y);
+
+  // interpolate base_keys at query_keys
+  const auto diff_x_vec = getSplineInterpolatedDiffValues(base_s, base_s, multi_spline_coef_x);
+  const auto diff_y_vec = getSplineInterpolatedDiffValues(base_s, base_s, multi_spline_coef_y);
+
+  std::vector<double> yaw_angle_vec;
+  for (size_t i = 0; i < diff_x_vec.size(); ++i) {
+    const double diff_x = diff_x_vec.at(i);
+    const double diff_y = diff_y_vec.at(i);
+
+    const double yaw = std::atan2(diff_y, diff_x);
+    yaw_angle_vec.push_back(yaw);
+  }
+
+  return yaw_angle_vec;
+}
 }  // namespace interpolation
+
+// member functions of SplineInterpolation1d
+void SplineInterpolation1d::calcSplineCoefficients(
+  const std::vector<double> & base_keys, const std::vector<double> & base_values)
+{
+  base_keys_ = base_keys;
+  multi_spline_coef_ = ::getSplineCoefficients(base_keys, base_values);
+}
+
+std::vector<double> SplineInterpolation1d::getSplineInterpolatedValues(
+  const std::vector<double> & query_keys) const
+{
+  return ::getSplineInterpolatedValues(base_keys_, query_keys, multi_spline_coef_);
+}
+
+// member functions of SplineInterpolationPoint2d
+// template <typename T>
+// void SplineInterpolationPoint2d::calcSplineCoefficients(const std::vector<T> & points)
+void SplineInterpolationPoint2d::calcSplineCoefficients(
+  const std::vector<geometry_msgs::msg::Point> & points)
+{
+  const auto base = getBaseValues(points);
+
+  base_s_vec_ = base.at(0);
+  const auto & base_x_vec = base.at(1);
+  const auto & base_y_vec = base.at(2);
+
+  // calculate spline coefficients
+  multi_spline_coef_x_ = getSplineCoefficients(base_s_vec_, base_x_vec);
+  multi_spline_coef_y_ = getSplineCoefficients(base_s_vec_, base_y_vec);
+}
+
+geometry_msgs::msg::Point SplineInterpolationPoint2d::getSplineInterpolatedPoint(
+  const size_t idx, const double s) const
+{
+  if (base_s_vec_.size() <= idx) {
+    throw std::out_of_range("idx is out of range.");
+  }
+
+  double whole_s = base_s_vec_.at(idx) + s;
+  if (whole_s < base_s_vec_.front()) {
+    whole_s = base_s_vec_.front();
+  }
+  if (whole_s > base_s_vec_.back()) {
+    whole_s = base_s_vec_.back();
+  }
```

Comment:
(@taikitanaka3 Related to this comment https://github.com/autowarefoundation/autoware.universe/pull/352#discussion_r801190638)

I removed throwing out of range here since it is hard to assure that `base_s_vec_.at(idx) + s` is within interpolation or not on the application side.

Reply:
FYI, previous code is
https://github.com/autowarefoundation/autoware.universe/blob/f73fe114085dd7972215a3f448800b81c5894812/common/interpolation/src/spline_interpolation.cpp#L331-L337